### PR TITLE
CORE-17025 / CORE-17413: Remove Hibernate & Introduce Lifecycle

### PR DESCRIPTION
**Pull request by jujoramos**
- Remove Hibernate code and dependencies from the State Manager.
- Make the State Manager Lifecycle aware so that it can be followed by
  higher components in the hierarchy.

### CODE REVIEW

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt [hunk 1]
```diff
@@ -1,27 +1,90 @@
...
+        )
+
+        entityManagerFactory = entityManagerFactoryProvider.create(
+            "corda-state-manager",
+            StateManagerEntities.classes,
+            DbEntityManagerConfiguration(dataSource)
+        )
+
+        eventHandler = CheckConnectionEventHandler(name) { dataSource.connection.close() }
+        lifecycleCoordinator = lifecycleCoordinatorFactory.createCoordinator(name, eventHandler)
```
Comment (sprintengine) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: Wonder should we add some info logs to say we've created or closed a StateManager

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/maintenance/FlowMaintenanceImplTests.kt [hunk 2]
```diff
@@ -23,37 +22,62 @@ import org.mockito.kotlin.never
...
     private val subscriptionFactory = mock<SubscriptionFactory> {
         on { createDurableSubscription(any(), any<SessionTimeoutTaskProcessor>(), any(), any()) } doReturn(subscription)
     }
+
     private val stateManager = mock<StateManager>()
     private val stateManagerFactory = mock<StateManagerFactory> {
         on { create(any()) } doReturn (stateManager)
     }
 
+    private val lifecycleCoordinator = mock<LifecycleCoordinator> {
```
Comment (JamesHR3) [inline components/flow/flow-service/src/test/kotlin/net/corda/flow/maintenance/FlowMaintenanceImplTests.kt]: I think the lifecycle library provides a test coordinator that might help you do this. It was designed more to test component lifecycle so it might not quite help with managed resources, but worth looking into.

File: libs/state-manager/state-manager-api/build.gradle [hunk 3]
```diff
@@ -1,13 +1,15 @@
 plugins {
-    id 'corda.common-publishing'
     id 'corda.common-library'
+    id 'corda.common-publishing'
```
Comment (JamesHR3) [inline libs/state-manager/state-manager-api/build.gradle]: Is the diff in this file necessary? I'm not quite sure what the reordering of the plugins or API modules is doing.

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt [hunk 4]
```diff
@@ -1,27 +1,90 @@
...
+    private val eventHandler : LifecycleEventHandler
+    private val lifecycleCoordinator: LifecycleCoordinator
+    override val name = LifecycleCoordinatorName("StateManager", UUID.randomUUID().toString())
 
     private companion object {
         private val objectMapper = ObjectMapper()
         private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
     }
 
+    init {
```
Comment (sprintengine) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: Personally I'd prefer this init logic to be in the `StateManagerFactory` as now `StateManager` has to be concerned with creating datasources, EMFs, lifecycle coordinators, etc. Feels more like a factory's job. But I don't want to hold up the PR for it

File: libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt [hunk 5]
```diff
@@ -1,11 +1,21 @@
...
 
+import net.corda.lifecycle.Lifecycle
+import net.corda.lifecycle.LifecycleCoordinatorName
 import net.corda.lifecycle.Resource
 
 /**
  * The [StateManager] provides functions to manage states within the underlying persistent storage.
  */
-interface StateManager : Resource {
+interface StateManager : Lifecycle, Resource {
```
Comment (JamesHR3) [inline libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt]: Not something to address here necessarily as I think subscriptions suffer the same thing, but I think things being both `Lifecycle` and `Resource` is a bit of an anti-pattern. `Lifecycle` is supposed to be about self managing the lifecycle of the component, while `Resource` is supposed to be about signaling that it is owned by something else.

File: .github/CODEOWNERS [hunk 6]
```diff
@@ -1,107 +1 @@
...
-/libs/uniqueness/                                               @corda/notaries
-/notary-plugins/                                                @corda/notaries
-/processors/uniqueness-processor/                               @corda/notaries
-/testing/uniqueness/                                            @corda/notaries
-
-# Ledger token selection files to be reviewed by the REST team
-# This needs to be after the ledger rules to partially override those
-/components/ledger/ledger-utxo-token-cache                      @corda/rest
-/components/ledger/ledger-utxo-flow/src/main/kotlin/net/corda/ledger/utxo/impl/token @corda/rest
+* @driessamyn @jasonbyrner3 @dimosr @ronanbrowne @rick-r3 @simon-johnson-r3 @blsemo @Omar-awad @aditisdesai @vinir3 @vkolomeyko @thiagoviana @Sakpal
```
Comment (JamesHR3) [inline .github/CODEOWNERS]: You'll probably need to revert this change after the retarget.

File: libs/membership/membership-impl/src/test/kotlin/net/corda/membership/lib/impl/grouppolicy/GroupPolicyParserImplTest.kt [hunk 7]
```diff
@@ -103,6 +105,17 @@ class GroupPolicyParserImplTest {
         assertThrows<BadGroupPolicyException> { groupPolicyParser.parse(holdingIdentity, INVALID_FORMAT_GROUP_POLICY) { null } }
     }
 
+    @Test
+    fun `Duplicate keys in group policy throws corda runtime exception`() {
```
Comment (JamesHR3) [inline libs/membership/membership-impl/src/test/kotlin/net/corda/membership/lib/impl/grouppolicy/GroupPolicyParserImplTest.kt]: Has this come in from the retarget?

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt [hunk 8]
```diff
@@ -1,27 +1,90 @@
...
+            validationTimeout = validationTimeout
+        )
+
+        entityManagerFactory = entityManagerFactoryProvider.create(
+            "corda-state-manager",
+            StateManagerEntities.classes,
+            DbEntityManagerConfiguration(dataSource)
+        )
+
+        eventHandler = CheckConnectionEventHandler(name) { dataSource.connection.close() }
```
Comment (sprintengine) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: Is the connection check closing the datasource connection? I see it's done in [DbConnectionManager](https://github.com/jujoramos/corda-runtime-os/blob/eaee01430026c6d7ab77f547a4cbdc5445e93a16/components/db/db-connection-manager-impl/src/main/kotlin/net/corda/db/connection/manager/impl/DbConnectionManagerImpl.kt#L168) - is that right?

File: libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt [hunk 9]
```diff
@@ -1,11 +1,21 @@
...
+
+    /**
+     * The State Manager lifecycle coordinator identifier.
+     *
+     * There might be multiple State Manager instances within a single process, the [name] should identify each one
+     * uniquely so the [Lifecycle] library can be used to follow regular component events.
+     */
+    val name: LifecycleCoordinatorName
```
Comment (sprintengine) [inline libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt:17]: I'm not sure what we'll use `name` for, does it need to be on this API?

File: libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt (hunk 5 above)
Comment (sprintengine) [inline libs/state-manager/state-manager-api/src/main/kotlin/net/corda/libs/statemanager/api/StateManager.kt]: I wonder do we actually need these interfaces extended by `StateManager` or should they just be implemented by the `StateManagerImpl` 

EDIT - `Resource` allows us to create a managed resource, so yes for that one.

File: testing/db-testkit/src/main/kotlin/net/corda/db/testkit/DbUtils.kt [hunk 10]
```diff
@@ -135,6 +139,59 @@ object DbUtils {
         }
     }
 
+    fun createStateManagerConfig(
```
Comment (JamesHR3) [inline testing/db-testkit/src/main/kotlin/net/corda/db/testkit/DbUtils.kt]: The DBUtils stuff is used to do in-memory testing in some circumstances (integration tests mainly). Does this config work for that? My understanding is that there's a problem with the state manager stuff that means it won't work in-memory, which may cause us some problems integrating this.

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+
+class CheckConnectionEventHandler(
+    private val componentName: LifecycleCoordinatorName,
+    private val connectionCheck: () -> Unit
+) : LifecycleEventHandler {
+
+    companion object {
+        private val logger = LoggerFactory.getLogger(this::class.java.enclosingClass)
+        val interval = 1.minutes
+        const val CHECK_EVENT_KEY = "CheckConnection"
```
Comment (JamesHR3) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt:21]: Should these be private?

File: components/flow/flow-service/src/main/kotlin/net/corda/flow/maintenance/FlowMaintenanceImpl.kt [hunk 12]
```diff
@@ -83,7 +82,6 @@ class FlowMaintenanceImpl @Activate constructor(
         when (event) {
             is StartEvent -> {
                 coordinator.updateStatus(LifecycleStatus.UP)
-                // TODO - this should register to follow the State Manager's lifecycle
```
Comment (JamesHR3) [inline components/flow/flow-service/src/main/kotlin/net/corda/flow/maintenance/FlowMaintenanceImpl.kt:110]: This is where the `Lifecycle` vs `Resource` thing is a problem, as it's not clear how to set up the lifecycle dependency properly for resources. I think we should solve for that separately.

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt (hunk 8 above)
Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: It retrieves a connection from the pool and then closes it, if the connection can not be obtained an exception is thrown (that's where the check "fails" and the component goes to `DOWN` state).

File: libs/state-manager/state-manager-api/build.gradle (hunk 3 above)
Comment (jujoramos) [reply libs/state-manager/state-manager-api/build.gradle]: Will revert, thanks 👍

File: .github/CODEOWNERS (hunk 6 above)
Comment (jujoramos) [reply .github/CODEOWNERS]: Done 👍

File: libs/membership/membership-impl/src/test/kotlin/net/corda/membership/lib/impl/grouppolicy/GroupPolicyParserImplTest.kt (hunk 7 above)
Comment (jujoramos) [reply libs/membership/membership-impl/src/test/kotlin/net/corda/membership/lib/impl/grouppolicy/GroupPolicyParserImplTest.kt]: Yeah, reverted 👍

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt (hunk 11 above)
Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt:21]: They are used from within `CheckConnectionEventHandlerTest`, that's why I didn't make them `private`.

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt (hunk 1 above)
Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: It's being logged as part of the `CheckConnectionEventHandler`:

```
2023-10-16 11:37:35.304 [lifecycle-coordinator-16] INFO  net.corda.libs.statemanager.impl.lifecycle.CheckConnectionEventHandler {} - StateManager-a76ca03e-1ab8-41e2-9927-9ee7972c6689 is starting
2023-10-16 11:37:35.328 [lifecycle-coordinator-16] INFO  net.corda.lifecycle.impl.LifecycleProcessor {} - Updating coordinator StateManager-a76ca03e-1ab8-41e2-9927-9ee7972c6689 from status DOWN to UP. Reason: Connection check passed
2023-10-16 11:39:00.877 [lifecycle-coordinator-70] INFO  net.corda.lifecycle.impl.LifecycleProcessor {} - Updating coordinator StateManager-a76ca03e-1ab8-41e2-9927-9ee7972c6689 from status UP to DOWN. Reason: Component has been stopped
2023-10-16 11:39:00.878 [lifecycle-coordinator-70] INFO  net.corda.libs.statemanager.impl.lifecycle.CheckConnectionEventHandler {} - StateManager-a76ca03e-1ab8-41e2-9927-9ee7972c6689 is stopping
```

File: components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/FlowMapperService.kt [hunk 13]
```diff
@@ -110,6 +110,8 @@ class FlowMapperService @Activate constructor(
 
             val stateManager = coordinator.createManagedResource(STATE_MANAGER) {
                 stateManagerFactory.create(stateManagerConfig)
+            }.also {
+                it.start()
             }
```
Comment (sprintengine) [inline components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/FlowMapperService.kt]: Think we also need to update this in FlowExecutorImpl where the stateManager managed resource is commented out.

https://github.com/jujoramos/corda-runtime-os/blob/36e96e5175385d9bf7eccc4e142a28a555b91b99/components/flow/flow-service/src/main/kotlin/net/corda/flow/service/FlowExecutorImpl.kt#L77-L80

Comment (jujoramos) [reply components/flow/flow-mapper-service/src/main/kotlin/net/corda/session/mapper/service/FlowMapperService.kt]: 👍

File: components/flow/flow-service/src/test/kotlin/net/corda/flow/maintenance/FlowMaintenanceImplTests.kt (hunk 2 above)
Comment (jujoramos) [reply components/flow/flow-service/src/test/kotlin/net/corda/flow/maintenance/FlowMaintenanceImplTests.kt]: I don't think it helps in this case, and I just need a mock implementation here so we should be fine keeping as it is (let me know if you don't agree, though).

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt (hunk 4 above)
Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/StateManagerImpl.kt]: Reverted 👍

File: testing/db-testkit/src/main/kotlin/net/corda/db/testkit/DbUtils.kt (hunk 10 above)
Comment (jujoramos) [reply testing/db-testkit/src/main/kotlin/net/corda/db/testkit/DbUtils.kt]: Yeah, I added this for the postgres part only and left the other section as a placeholder for when we make the state manager compatible with in-memory database. It has been reverted as we don't need it anymore 👍

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt [hunk 14]
```diff
@@ -0,0 +1,57 @@
...
+    }
+
+    override fun processEvent(event: LifecycleEvent, coordinator: LifecycleCoordinator) {
+        logger.debug("{} processing event {}", componentName, event)
+
+        when (event) {
+            is StartEvent -> {
+                logger.info("$componentName is starting")
+                scheduleConnectionCheck(coordinator, 0)
```
Comment (owenstanford) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt:30]: should this be overriding the default 1 min interval?

Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/lifecycle/CheckConnectionEventHandler.kt:30]: I think so, yeah, we want to check the connection straight away.

File: libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/repository/impl/PostgresQueryProvider.kt [hunk 15]
```diff
@@ -55,7 +59,7 @@ class PostgresQueryProvider : AbstractQueryProvider() {
         filters.map { "(${metadataKeyFilter(it)})" }
 
     fun metadataKeyFilter(filter: MetadataFilter) =
-        "(s.metadata->>'${filter.key}')::::${filter.value.toNativeType()} ${filter.operation.toNativeOperator()} '${filter.value}'"
+        "(s.$METADATA_COLUMN->>'${filter.key}')::${filter.value.toNativeType()} ${filter.operation.toNativeOperator()} '${filter.value}'"
```
Comment (sprintengine) [inline libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/repository/impl/PostgresQueryProvider.kt:62]: Was this extra "::" causing a bug before?

Comment (jujoramos) [reply libs/state-manager/state-manager-db-impl/src/main/kotlin/net/corda/libs/statemanager/impl/repository/impl/PostgresQueryProvider.kt:62]: No bug, nop, it was the syntax I had to use when using hibernate. Now that we got rid of it, we can write the query as we'd do when interacting with the database directly.

**Review by conal-smith — APPROVED** [review]
LGTM

**Review by Omar-awad — APPROVED** [review]
LGTM

**Review by JamesHR3 — APPROVED** [review]
